File: UsingPython/quicksort.py
"""
implement the standard quicksort algorithm
"""

def partition(l, low, high):
    pivot = l[high]

    i=low
    j=high

    """ what is the invariant we are maintaining
    it is:
        values < pivot are in range [low, i)
        values > pivot are in range (j, high]
    so the point where we stop partitioning is when i==j
    """
    while i < j:
        if l[i] < pivot:
            i += 1
        else:
            j -= 1
            l[i], l[j] = l[j], l[i]
        #print 'i: {}, j: {}'.format(i,j)
    l[j], l[high] = l[high], l[j]
    return j #pivot index

def quicksort(l, low, high):

    if (low >= high):
        return

    pivot = partition(l, low, high)
    #print 'low:{}, high: {}, pivotIndex:{}, list: {}'.format(low, high, pivot, l)
    quicksort(l, low, pivot-1)
    quicksort(l, pivot+1, high)



def main():
    import random
    l = random.sample(xrange(1000), 10)

    #print 'unsorted: {}'.format(l)
    quicksort(l, 0, len(l)-1)
    #print 'sorted: {}'.format(l)


if __name__ == "__main__":
    main()

File: CCI/SortingSearching/quickSort.cc
#include <iostream>
#include <vector>
#include <random>

using std::vector;
using std::cout;
using std::endl;
using std::swap;

int partition(vector<int>& array, int low, int high)
{
  int pivotIndex = high;
  int pivot = array[high];

  int i = low, j = high-1;

  while(i <= j) {
    if (array[i] <= pivot)
    {
      ++i;
    }
    else
    {
      swap(array[i], array[j]);
      --j;
    }
  }

  swap(array[pivotIndex], array[i]);
  return i;
}

void quickSortHelper(vector<int>& array, int low, int high)
{
  if (low >= high)
    return;

  int pivotIndex = partition(array, low, high);
  quickSortHelper(array, low, pivotIndex-1);
  quickSortHelper(array, pivotIndex+1, high);
}

void quickSort(vector<int>& array)
{
  quickSortHelper(array, 0, array.size());
}

void printArray(const vector<int>& array)
{
  for (auto a: array)
    cout << a << ",";
  cout << endl;
}

int main()
{
  vector<int> array = {1,2,1,4,5,6,1,8,9,10};
  random_shuffle(array.begin(), array.end());
  printArray(array);
  quickSort(array);
  printArray(array);

  return 0;
}
